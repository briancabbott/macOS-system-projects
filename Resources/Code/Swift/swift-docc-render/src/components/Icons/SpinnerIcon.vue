<!--
  This source file is part of the Swift.org open source project

  Copyright (c) 2022 Apple Inc. and the Swift project authors
  Licensed under Apache License v2.0 with Runtime Library Exception

  See https://swift.org/LICENSE.txt for license information
  See https://swift.org/CONTRIBUTORS.txt for Swift project authors
-->

<template>
  <SVGIcon class="spinner-icon" viewBox="0 0 39.02 39.02" themeId="spinner">
    <path d="m15.529 11.96-3.57 3.569-7.99-7.99 3.57-3.57z"/>
    <path d="M0 22.072v-5.06h11.331v5.06z"/>
    <path d="m7.54 35.096-3.57-3.569 7.99-7.99 3.57 3.569z"/>
    <path d="M22.057 39.02H17.01v-11.3h5.047z"/>
    <path d="m35.096 31.528-3.569 3.568-7.99-7.99 3.569-3.569z"/>
    <path d="M39.02 17.01v5.046h-11.3V17.01z"/>
    <path d="m31.528 3.97 3.569 3.57-7.99 7.99-3.57-3.57z"/>
    <path d="M17.011 0h5.061v11.331h-5.061z"/>
  </SVGIcon>
</template>

<script>
import SVGIcon from 'docc-render/components/SVGIcon.vue';

export default {
  name: 'SpinnerIcon',
  components: { SVGIcon },
};
</script>

<style scoped lang="scss">
@keyframes fadeout {
  from {
    opacity: 1;
  }

  to {
    opacity: 0;
  }
}

path {
  $num-rects: 8;
  $animation-duration-full: 1000ms;
  $animation-duration-per-rect: $animation-duration-full / $num-rects;

  animation-duration: $animation-duration-full;
  animation-iteration-count: infinite;
  animation-name: fadeout;
  fill: currentColor;

  @for $i from 1 through $num-rects {
    &:nth-of-type(#{$i}) {
      $animation-delay: ($i - 1) * $animation-duration-per-rect * -1;
      animation-delay: $animation-delay;
    }
  }
}
</style>
