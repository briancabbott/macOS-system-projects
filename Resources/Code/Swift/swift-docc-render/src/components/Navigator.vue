<!--
  This source file is part of the Swift.org open source project

  Copyright (c) 2022 Apple Inc. and the Swift project authors
  Licensed under Apache License v2.0 with Runtime Library Exception

  See https://swift.org/LICENSE.txt for license information
  See https://swift.org/CONTRIBUTORS.txt for Swift project authors
-->

<template>
  <nav
    :aria-labelledby="INDEX_ROOT_KEY"
    class="navigator"
  >
    <NavigatorCard
      v-if="!isFetching"
      :technology="technology.title"
      :is-technology-beta="technology.beta"
      :technology-path="technology.path || technology.url"
      :type="type"
      :children="flatChildren"
      :active-path="activePath"
      :scrollLockID="scrollLockID"
      :error-fetching="errorFetching"
      :render-filter-on-top="renderFilterOnTop"
      :api-changes="apiChanges"
      :allow-hiding="allowHiding"
      :enableQuickNavigation="enableQuickNavigation"
      :navigator-references="navigatorReferences"
      @close="$emit('close')"
    />
    <NavigatorCardInner v-else class="loading-placeholder">
      <transition name="delay-visibility" appear>
        <SpinnerIcon class="loading-spinner" />
      </transition>
    </NavigatorCardInner>
    <div aria-live="polite" class="visuallyhidden">
      Navigator is {{ isFetching ? 'loading' : 'ready' }}
    </div>
  </nav>
</template>

<script>
import QuickNavigationStore from 'docc-render/stores/QuickNavigationStore';
import NavigatorCard from 'theme/components/Navigator/NavigatorCard.vue';
import SpinnerIcon from 'theme/components/Icons/SpinnerIcon.vue';
import NavigatorCardInner from 'docc-render/components/Navigator/NavigatorCardInner.vue';
import { INDEX_ROOT_KEY } from 'docc-render/constants/sidebar';
import { TopicTypes } from 'docc-render/constants/TopicTypes';
import { getSetting } from 'docc-render/utils/theme-settings';

/**
 * @typedef NavigatorFlatItem
 * @property {number} uid - generated UID
 * @property {string} title - title of symbol
 * @property {string} type - symbol type, used for the icon
 * @property {string} icon - an image reference to override the type icon
 * @property {array} abstract - symbol abstract
 * @property {string} path - path to page, used in navigation
 * @property {number} parent - parent UID
 * @property {number} groupMarkerUID - UID of the groupMarker that labels this
 * @property {number} deprecatedChildrenCount - number of children that are deprecated.
 * Used for filtering
 * @property {number} depth - depth of symbol in original tree
 * @property {number} index - index of item in siblings
 * @property {number} siblingsCount - number of siblings
 * @property {number[]} childUIDs - array of child UIDs
 * @property {boolean} deprecated - symbol is deprecated or is not
 */

/**
 * Renders a sidebar navigator component.
 */
export default {
  name: 'Navigator',
  components: {
    NavigatorCard,
    NavigatorCardInner,
    SpinnerIcon,
  },
  data() {
    return {
      INDEX_ROOT_KEY,
      store: QuickNavigationStore,
    };
  },
  props: {
    parentTopicIdentifiers: {
      type: Array,
      required: true,
    },
    technology: {
      type: Object,
      required: true,
    },
    isFetching: {
      type: Boolean,
      default: false,
    },
    references: {
      type: Object,
      default: () => {},
    },
    navigatorReferences: {
      type: Object,
      default: () => {},
    },
    scrollLockID: {
      type: String,
      default: '',
    },
    errorFetching: {
      type: Boolean,
      default: false,
    },
    renderFilterOnTop: {
      type: Boolean,
      default: false,
    },
    apiChanges: {
      type: Object,
      default: null,
    },
    allowHiding: {
      type: Boolean,
      default: true,
    },
  },
  provide() {
    return { store: this.store };
  },
  computed: {
    // gets the paths for each parent in the breadcrumbs
    parentTopicReferences({ references, parentTopicIdentifiers }) {
      return parentTopicIdentifiers
        .reduce((all, identifier) => {
          const reference = references[identifier];
          if (reference) return all.concat(reference);
          console.error(`Reference for "${identifier}" is missing`);
          return all;
        }, []);
    },
    // splits out the top-level technology crumb
    activePath({ parentTopicReferences, $route: { path } }) {
      // Ensure the path does not have a trailing slash
      // eslint-disable-next-line no-param-reassign
      path = path.replace(/\/$/, '').toLowerCase();
      // route's path is activePath on root
      if (!parentTopicReferences.length) return [path];
      let itemsToSlice = 1;
      // if the first item is a `technology`, slice off it and the technology itself
      if (parentTopicReferences[0].kind === 'technologies') {
        itemsToSlice = 2;
      }
      return parentTopicReferences.slice(itemsToSlice).map(r => r.url).concat(path);
    },
    enableQuickNavigation: () => (
      getSetting(['features', 'docs', 'quickNavigation', 'enable'], false)
    ),
    /**
     * Recomputes the list of flat children.
     * @return NavigatorFlatItem[]
     */
    flatChildren: ({
      enableQuickNavigation, flattenNestedData, technology = {}, store,
    }) => {
      const flatIndex = flattenNestedData(technology.children || [], null, 0, technology.beta);
      if (enableQuickNavigation) {
        store.setFlattenIndex(flatIndex);
      }
      return flatIndex;
    },
    /**
     * The root item is always a module
     */
    type: () => TopicTypes.module,
  },
  methods: {
    /**
     * Generates a unique hash, from a string, generating a signed number.
     * @returns Number
     */
    hashCode(str) {
      return str.split('').reduce((prevHash, currVal) => (
        // eslint-disable-next-line no-bitwise
        (((prevHash << 5) - prevHash) + currVal.charCodeAt(0)) | 0
      ), 0);
    },
    /**
     * @param {{path: string, type: string, title: string, children?: [] }[]} childrenNodes
     * @param {NavigatorFlatItem | null} parent
     * @param {Number} depth
     * @param {Boolean} parentBetaStatus
     * @return {NavigatorFlatItem[]}
     */
    flattenNestedData(childrenNodes, parent = null, depth = 0, parentBetaStatus = false) {
      let items = [];
      const len = childrenNodes.length;
      let index;
      // reference to the last label node
      let groupMarkerNode = null;
      for (index = 0; index < len; index += 1) {
        // get the children
        const { children, ...node } = childrenNodes[index];
        // generate the extra properties
        const { uid: parentUID = INDEX_ROOT_KEY } = parent || {};
        // generate a uid to track by
        node.uid = this.hashCode(`${parentUID}+${node.path}_${depth}_${index}`);
        // store the parent uid
        node.parent = parentUID;
        // store the current groupMarker reference
        if (node.type === TopicTypes.groupMarker) {
          node.deprecatedChildrenCount = 0;
          groupMarkerNode = node;
        } else if (groupMarkerNode) {
          // push the current node to the group marker before it
          groupMarkerNode.childUIDs.push(node.uid);
          // store the groupMarker UID for each item
          node.groupMarkerUID = groupMarkerNode.uid;
          if (node.deprecated) {
            // count deprecated children, so we can hide the entire group when filtering
            groupMarkerNode.deprecatedChildrenCount += 1;
          }
        }
        // store which item it is
        node.index = index;
        // store how many siblings it has
        node.siblingsCount = len;
        // store the depth
        node.depth = depth;
        // store child UIDs
        node.childUIDs = [];
        // if the parent is not the root, push to its childUIDs the current node uid
        if (parent) {
          // push child to parent
          parent.childUIDs.push(node.uid);
        }
        // if the parent or the entire technology are marked as `Beta`,
        // child elements do not get marked as `Beta`.
        if (node.beta && parentBetaStatus) {
          node.beta = false;
        }

        items.push(node);
        if (children) {
          // return the children to the parent
          items = items.concat(this.flattenNestedData(
            children, node, depth + 1, parentBetaStatus || node.beta,
          ));
        }
      }
      return items;
    },
  },
};
</script>

<style scoped lang='scss'>
@import 'docc-render/styles/_core.scss';

.navigator {
  height: 100%;
  display: flex;
  flex-flow: column;

  @include breakpoint(medium, nav) {
    position: static;
    transition: none;
  }
}

.loading-placeholder {
  align-items: center;
  color: var(--color-figure-gray-secondary);
  justify-content: center;
  height: 100vh;
  position: sticky;
  top: var(--nav-height, 0px);
}

.loading-spinner {
  --spinner-size: 40px; // used for both width and height
  --spinner-delay: 1s; // don't show spinner until this much time has passed

  height: var(--spinner-size);
  width: var(--spinner-size);

  &.delay-visibility-enter-active {
    transition: visibility var(--spinner-delay);
    visibility: hidden;
  }
}
</style>
