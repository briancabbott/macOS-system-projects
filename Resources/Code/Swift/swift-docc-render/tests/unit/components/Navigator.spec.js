/**
 * This source file is part of the Swift.org open source project
 *
 * Copyright (c) 2021-2022 Apple Inc. and the Swift project authors
 * Licensed under Apache License v2.0 with Runtime Library Exception
 *
 * See https://swift.org/LICENSE.txt for license information
 * See https://swift.org/CONTRIBUTORS.txt for Swift project authors
*/

import Navigator from '@/components/Navigator.vue';
import { shallowMount } from '@vue/test-utils';
import NavigatorCard from '@/components/Navigator/NavigatorCard.vue';
import SpinnerIcon from '@/components/Icons/SpinnerIcon.vue';
import { baseNavStickyAnchorId } from 'docc-render/constants/nav';
import { TopicTypes } from '@/constants/TopicTypes';
import { INDEX_ROOT_KEY } from '@/constants/sidebar';
import { clone } from '@/utils/data';

jest.mock('docc-render/utils/throttle', () => jest.fn(v => v));

const technology = {
  title: 'FooTechnology',
  children: [
    {
      title: 'Group Marker',
      type: TopicTypes.groupMarker,
    },
    {
      title: 'Child0',
      path: '/foo/child0',
      type: 'article',
      children: [
        {
          title: 'Group Marker, Child 0',
          type: TopicTypes.groupMarker,
        },
        {
          title: 'Child0_GrandChild0',
          path: '/foo/child0/grandchild0',
          type: 'tutorial',
        },
        {
          title: 'Child0_GrandChild1',
          path: '/foo/child0/grandchild1',
          type: 'tutorial',
          children: [{
            title: 'Child0_GrandChild0_GreatGrandChild0',
            path: '/foo/child0/grandchild0/greatgrandchild0',
            type: 'tutorial',
          }],
        },
        {
          title: 'Child0_GrandChild2',
          path: '/foo/child0/grandchild2',
          type: 'tutorial',
        },
      ],
    },
    {
      title: 'Child1',
      path: '/foo/child1/',
      type: 'tutorial',
      children: [{
        title: 'Child1_GrandChild0',
        path: '/foo/child1/grandchild0',
        type: 'method',
      }],
    },
  ],
  path: '/path/to/technology',
};

const references = {
  technologies: { kind: 'technologies' },
  root: { url: 'root' },
  first: { url: 'first' },
  second: { url: 'second' },
  third: { url: 'third' },
};

const parentTopicIdentifiers = [
  'root', 'first', 'second',
];

const fallbackTechnology = {
  title: 'FallbackTechnology',
  url: '/url/to/technology',
};

const mocks = {
  $route: {
    path: '/current/path',
  },
};

const navigatorReferences = {
  foo: {},
};

const defaultProps = {
  parentTopicIdentifiers,
  technology,
  references,
  scrollLockID: 'foo',
  renderFilterOnTop: false,
  navigatorReferences,
};

const fauxAnchor = document.createElement('DIV');
Object.defineProperty(fauxAnchor, 'offsetTop', { value: 0, writable: true });
fauxAnchor.id = baseNavStickyAnchorId;
document.body.appendChild(fauxAnchor);

const createWrapper = ({ propsData, ...others } = {}) => shallowMount(Navigator, {
  propsData: {
    ...defaultProps,
    ...propsData,
  },
  mocks,
  ...others,
});

const errorSpy = jest.spyOn(console, 'error').mockReturnValue('');

describe('Navigator', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders the Navigator', () => {
    const wrapper = createWrapper();
    const navigator = wrapper.find('.navigator');
    // assert navigator is a `nav`
    expect(navigator.is('nav')).toBe(true);
    expect(navigator.attributes()).toHaveProperty('aria-labelledby', INDEX_ROOT_KEY);
    // assert Navigator card is rendered
    expect(wrapper.find(NavigatorCard).props()).toEqual({
      activePath: [references.first.url, references.second.url, mocks.$route.path],
      // will assert in another test
      children: expect.any(Array),
      enableQuickNavigation: false,
      type: TopicTypes.module,
      technology: technology.title,
      technologyPath: technology.path,
      isTechnologyBeta: false,
      scrollLockID: defaultProps.scrollLockID,
      renderFilterOnTop: defaultProps.renderFilterOnTop,
      errorFetching: false,
      apiChanges: null,
      allowHiding: true,
      navigatorReferences,
    });
    expect(wrapper.find('.loading-placeholder').exists()).toBe(false);
  });

  it('renders a loading placeholder, if is fetching', () => {
    const wrapper = createWrapper({
      propsData: {
        isFetching: true,
      },
    });
    // assert Navigator card is rendered
    expect(wrapper.find(NavigatorCard).exists()).toBe(false);

    const placeholder = wrapper.find('.loading-placeholder');
    expect(placeholder.exists()).toBe(true);
    expect(placeholder.contains(SpinnerIcon)).toBe(true);
  });

  it('renders an aria live that tells VO users when navigator is loading', () => {
    const wrapper = createWrapper({
      propsData: {
        isFetching: true,
      },
    });
    expect(wrapper.find('[aria-live="polite"]').exists()).toBe(true);
    expect(wrapper.find('[aria-live="polite"]').text()).toBe('Navigator is loading');
  });

  it('renders an aria live that tells VO users when navigator is ready', () => {
    const wrapper = createWrapper();
    expect(wrapper.find('[aria-live="polite"]').exists()).toBe(true);
    expect(wrapper.find('[aria-live="polite"]').text()).toBe('Navigator is ready');
  });

  it('falls back to using the `technology.url` for the `technology-path`', () => {
    const wrapper = createWrapper({
      propsData: {
        technology: fallbackTechnology,
      },
    });
    expect(wrapper.find(NavigatorCard).props()).toEqual({
      activePath: [references.first.url, references.second.url, mocks.$route.path],
      // will assert in another test
      children: [],
      enableQuickNavigation: false,
      type: TopicTypes.module,
      technology: fallbackTechnology.title,
      technologyPath: fallbackTechnology.url,
      isTechnologyBeta: false,
      scrollLockID: defaultProps.scrollLockID,
      renderFilterOnTop: defaultProps.renderFilterOnTop,
      errorFetching: false,
      apiChanges: null,
      allowHiding: true,
      navigatorReferences,
    });
  });

  it('passes the errorFetching prop', () => {
    const wrapper = createWrapper({
      propsData: {
        errorFetching: true,
      },
    });
    expect(wrapper.find(NavigatorCard).props('errorFetching')).toBe(true);
  });

  it('strips out possible technology URLs from the activePath', () => {
    const wrapper = createWrapper({
      propsData: {
        parentTopicIdentifiers: ['technologies'].concat(parentTopicIdentifiers),
      },
    });
    expect(wrapper.find(NavigatorCard).props('activePath')).toEqual([
      references.first.url, references.second.url, mocks.$route.path,
    ]);
  });

  it('casts to lowercase and strips out trailing slashes from the last activePath item', () => {
    const wrapper = createWrapper({
      mocks: {
        ...mocks,
        $route: {
          ...mocks.$route,
          path: '/documentation/Foo/Bar/',
        },
      },
    });
    expect(wrapper.find(NavigatorCard).props('activePath')).toEqual([
      references.first.url, references.second.url, '/documentation/foo/bar',
    ]);
  });

  it('renders the root path as activePath when there is no parentTopicIdentifiers', () => {
    const wrapper = createWrapper({
      propsData: {
        parentTopicIdentifiers: [],
      },
    });
    expect(wrapper.find(NavigatorCard).props('activePath')).toEqual([mocks.$route.path]);
  });

  it('renders the root path as activePath when there are no valid parentTopicIdentifiers', () => {
    const identifier = 'ref://invalid';
    const wrapper = createWrapper({
      propsData: {
        parentTopicIdentifiers: [identifier],
      },
    });
    expect(wrapper.find(NavigatorCard).props('activePath')).toEqual([mocks.$route.path]);
    expect(errorSpy).toHaveBeenCalledTimes(1);
    expect(errorSpy).toHaveBeenCalledWith(`Reference for "${identifier}" is missing`);
  });

  it('removes any parent topic identifiers, which dont have a reference', () => {
    const wrapper = createWrapper({
      propsData: {
        references: {
          root: { url: 'root' },
          first: { url: 'first' },
          // skip the `second` reference
        },
      },
    });
    // assert `second` is missing from the activePath
    expect(wrapper.find(NavigatorCard).props('activePath')).toEqual([
      references.first.url, mocks.$route.path,
    ]);
    expect(errorSpy).toHaveBeenCalledTimes(1);
    expect(errorSpy).toHaveBeenCalledWith('Reference for "second" is missing');
  });

  it('re-emits the `@close` event', () => {
    const wrapper = createWrapper();
    wrapper.find(NavigatorCard).vm.$emit('close');
    expect(wrapper.emitted('close')).toHaveLength(1);
  });

  it('generates a unique hash', () => {
    const wrapper = createWrapper();
    // make sure hashCode returns a unique hash from a string
    expect(wrapper.vm.hashCode('foo')).toBe(101574);
    expect(wrapper.vm.hashCode('oof')).toBe(110214);
    expect(wrapper.vm.hashCode('ofo')).toBe(109944);
    expect(wrapper.vm.hashCode('foo1')).toBe(3148843);
    // make sure hash is never too big
    expect(wrapper.vm.hashCode('foo1'.repeat(10))).toBe(-1535108562);
    expect(wrapper.vm.hashCode('foo1'.repeat(50))).toBe(-2107259162);
  });

  it('flattens deeply nested children and provides them to the NavigatorCard', () => {
    const wrapper = createWrapper();
    expect(wrapper.find(NavigatorCard).props('children')).toEqual([
      {
        childUIDs: [
          551503844,
          -97593391,
        ],
        deprecatedChildrenCount: 0,
        depth: 0,
        index: 0,
        parent: INDEX_ROOT_KEY,
        siblingsCount: 3,
        title: 'Group Marker',
        type: 'groupMarker',
        uid: -196255993,
      },
      {
        childUIDs: [
          -361407047,
          1438225895,
          1439149417,
          1440072939,
        ],
        depth: 0,
        groupMarkerUID: -196255993,
        index: 1,
        parent: INDEX_ROOT_KEY,
        path: '/foo/child0',
        siblingsCount: 3,
        title: 'Child0',
        type: 'article',
        uid: 551503844,
      },
      {
        childUIDs: [
          1438225895,
          1439149417,
          1440072939,
        ],
        deprecatedChildrenCount: 0,
        depth: 1,
        index: 0,
        parent: 551503844,
        siblingsCount: 4,
        title: 'Group Marker, Child 0',
        type: 'groupMarker',
        uid: -361407047,
      },
      {
        childUIDs: [],
        depth: 1,
        groupMarkerUID: -361407047,
        index: 1,
        parent: 551503844,
        path: '/foo/child0/grandchild0',
        siblingsCount: 4,
        title: 'Child0_GrandChild0',
        type: 'tutorial',
        uid: 1438225895,
      },
      {
        childUIDs: [
          305326087,
        ],
        depth: 1,
        groupMarkerUID: -361407047,
        index: 2,
        parent: 551503844,
        path: '/foo/child0/grandchild1',
        siblingsCount: 4,
        title: 'Child0_GrandChild1',
        type: 'tutorial',
        uid: 1439149417,
      },
      {
        childUIDs: [],
        depth: 2,
        index: 0,
        parent: 1439149417,
        path: '/foo/child0/grandchild0/greatgrandchild0',
        siblingsCount: 1,
        title: 'Child0_GrandChild0_GreatGrandChild0',
        type: 'tutorial',
        uid: 305326087,
      },
      {
        childUIDs: [],
        depth: 1,
        groupMarkerUID: -361407047,
        index: 3,
        parent: 551503844,
        path: '/foo/child0/grandchild2',
        siblingsCount: 4,
        title: 'Child0_GrandChild2',
        type: 'tutorial',
        uid: 1440072939,
      },
      {
        childUIDs: [
          -827353283,
        ],
        depth: 0,
        groupMarkerUID: -196255993,
        index: 2,
        parent: INDEX_ROOT_KEY,
        path: '/foo/child1/',
        siblingsCount: 3,
        title: 'Child1',
        type: 'tutorial',
        uid: -97593391,
      },
      {
        childUIDs: [],
        depth: 1,
        index: 0,
        parent: -97593391,
        path: '/foo/child1/grandchild0',
        siblingsCount: 1,
        title: 'Child1_GrandChild0',
        type: 'method',
        uid: -827353283,
      },
    ]);
  });

  it('counts the amount of deprecated items a groupMarker has', () => {
    const technologyClone = clone(technology);
    technologyClone.children[1].deprecated = true;
    technologyClone.children[2].deprecated = true;
    technologyClone.children[1].children[1].deprecated = true;
    const wrapper = createWrapper({
      propsData: {
        technology: technologyClone,
      },
    });
    const children = wrapper.find(NavigatorCard).props('children');
    expect(children[0]).toHaveProperty('deprecatedChildrenCount', 2);
    expect(children).toMatchSnapshot();
  });

  it('removes the `beta` flag from children, if the technology is a `beta`', () => {
    const technologyClone = clone(technology);
    technologyClone.beta = true;
    technologyClone.children[1].beta = true;
    technologyClone.children[1].children[0].beta = true;
    const wrapper = createWrapper({
      propsData: {
        technology: technologyClone,
      },
    });
    expect(wrapper.find(NavigatorCard).props('children')).toMatchSnapshot();
  });

  it('removes the `beta` flag from children, if the parent is a `beta`', () => {
    const technologyClone = clone(technology);
    technologyClone.children[1].beta = true;
    technologyClone.children[1].children[1].beta = true;
    // case where the direct parent is NOT `Beta`, but an ancestor is
    technologyClone.children[1].children[2].children[0].beta = true;
    // set an end node as beta
    technologyClone.children[2].children[0].beta = true;
    const wrapper = createWrapper({
      propsData: {
        technology: technologyClone,
      },
    });
    expect(wrapper.find(NavigatorCard).props('children')).toMatchSnapshot();
  });
});
