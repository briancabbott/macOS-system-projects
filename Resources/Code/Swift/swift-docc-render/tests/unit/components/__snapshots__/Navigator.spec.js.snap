// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navigator counts the amount of deprecated items a groupMarker has 1`] = `
Array [
  Object {
    "childUIDs": Array [
      551503844,
      -97593391,
    ],
    "deprecatedChildrenCount": 2,
    "depth": 0,
    "index": 0,
    "parent": <root>,
    "siblingsCount": 3,
    "title": "Group Marker",
    "type": "groupMarker",
    "uid": -196255993,
  },
  Object {
    "childUIDs": Array [
      -361407047,
      1438225895,
      1439149417,
      1440072939,
    ],
    "deprecated": true,
    "depth": 0,
    "groupMarkerUID": -196255993,
    "index": 1,
    "parent": <root>,
    "path": "/foo/child0",
    "siblingsCount": 3,
    "title": "Child0",
    "type": "article",
    "uid": 551503844,
  },
  Object {
    "childUIDs": Array [
      1438225895,
      1439149417,
      1440072939,
    ],
    "deprecatedChildrenCount": 1,
    "depth": 1,
    "index": 0,
    "parent": 551503844,
    "siblingsCount": 4,
    "title": "Group Marker, Child 0",
    "type": "groupMarker",
    "uid": -361407047,
  },
  Object {
    "childUIDs": Array [],
    "deprecated": true,
    "depth": 1,
    "groupMarkerUID": -361407047,
    "index": 1,
    "parent": 551503844,
    "path": "/foo/child0/grandchild0",
    "siblingsCount": 4,
    "title": "Child0_GrandChild0",
    "type": "tutorial",
    "uid": 1438225895,
  },
  Object {
    "childUIDs": Array [
      305326087,
    ],
    "depth": 1,
    "groupMarkerUID": -361407047,
    "index": 2,
    "parent": 551503844,
    "path": "/foo/child0/grandchild1",
    "siblingsCount": 4,
    "title": "Child0_GrandChild1",
    "type": "tutorial",
    "uid": 1439149417,
  },
  Object {
    "childUIDs": Array [],
    "depth": 2,
    "index": 0,
    "parent": 1439149417,
    "path": "/foo/child0/grandchild0/greatgrandchild0",
    "siblingsCount": 1,
    "title": "Child0_GrandChild0_GreatGrandChild0",
    "type": "tutorial",
    "uid": 305326087,
  },
  Object {
    "childUIDs": Array [],
    "depth": 1,
    "groupMarkerUID": -361407047,
    "index": 3,
    "parent": 551503844,
    "path": "/foo/child0/grandchild2",
    "siblingsCount": 4,
    "title": "Child0_GrandChild2",
    "type": "tutorial",
    "uid": 1440072939,
  },
  Object {
    "childUIDs": Array [
      -827353283,
    ],
    "deprecated": true,
    "depth": 0,
    "groupMarkerUID": -196255993,
    "index": 2,
    "parent": <root>,
    "path": "/foo/child1/",
    "siblingsCount": 3,
    "title": "Child1",
    "type": "tutorial",
    "uid": -97593391,
  },
  Object {
    "childUIDs": Array [],
    "depth": 1,
    "index": 0,
    "parent": -97593391,
    "path": "/foo/child1/grandchild0",
    "siblingsCount": 1,
    "title": "Child1_GrandChild0",
    "type": "method",
    "uid": -827353283,
  },
]
`;

exports[`Navigator removes the \`beta\` flag from children, if the parent is a \`beta\` 1`] = `
Array [
  Object {
    "childUIDs": Array [
      551503844,
      -97593391,
    ],
    "deprecatedChildrenCount": 0,
    "depth": 0,
    "index": 0,
    "parent": <root>,
    "siblingsCount": 3,
    "title": "Group Marker",
    "type": "groupMarker",
    "uid": -196255993,
  },
  Object {
    "beta": true,
    "childUIDs": Array [
      -361407047,
      1438225895,
      1439149417,
      1440072939,
    ],
    "depth": 0,
    "groupMarkerUID": -196255993,
    "index": 1,
    "parent": <root>,
    "path": "/foo/child0",
    "siblingsCount": 3,
    "title": "Child0",
    "type": "article",
    "uid": 551503844,
  },
  Object {
    "childUIDs": Array [
      1438225895,
      1439149417,
      1440072939,
    ],
    "deprecatedChildrenCount": 0,
    "depth": 1,
    "index": 0,
    "parent": 551503844,
    "siblingsCount": 4,
    "title": "Group Marker, Child 0",
    "type": "groupMarker",
    "uid": -361407047,
  },
  Object {
    "beta": false,
    "childUIDs": Array [],
    "depth": 1,
    "groupMarkerUID": -361407047,
    "index": 1,
    "parent": 551503844,
    "path": "/foo/child0/grandchild0",
    "siblingsCount": 4,
    "title": "Child0_GrandChild0",
    "type": "tutorial",
    "uid": 1438225895,
  },
  Object {
    "childUIDs": Array [
      305326087,
    ],
    "depth": 1,
    "groupMarkerUID": -361407047,
    "index": 2,
    "parent": 551503844,
    "path": "/foo/child0/grandchild1",
    "siblingsCount": 4,
    "title": "Child0_GrandChild1",
    "type": "tutorial",
    "uid": 1439149417,
  },
  Object {
    "beta": false,
    "childUIDs": Array [],
    "depth": 2,
    "index": 0,
    "parent": 1439149417,
    "path": "/foo/child0/grandchild0/greatgrandchild0",
    "siblingsCount": 1,
    "title": "Child0_GrandChild0_GreatGrandChild0",
    "type": "tutorial",
    "uid": 305326087,
  },
  Object {
    "childUIDs": Array [],
    "depth": 1,
    "groupMarkerUID": -361407047,
    "index": 3,
    "parent": 551503844,
    "path": "/foo/child0/grandchild2",
    "siblingsCount": 4,
    "title": "Child0_GrandChild2",
    "type": "tutorial",
    "uid": 1440072939,
  },
  Object {
    "childUIDs": Array [
      -827353283,
    ],
    "depth": 0,
    "groupMarkerUID": -196255993,
    "index": 2,
    "parent": <root>,
    "path": "/foo/child1/",
    "siblingsCount": 3,
    "title": "Child1",
    "type": "tutorial",
    "uid": -97593391,
  },
  Object {
    "beta": true,
    "childUIDs": Array [],
    "depth": 1,
    "index": 0,
    "parent": -97593391,
    "path": "/foo/child1/grandchild0",
    "siblingsCount": 1,
    "title": "Child1_GrandChild0",
    "type": "method",
    "uid": -827353283,
  },
]
`;

exports[`Navigator removes the \`beta\` flag from children, if the technology is a \`beta\` 1`] = `
Array [
  Object {
    "childUIDs": Array [
      551503844,
      -97593391,
    ],
    "deprecatedChildrenCount": 0,
    "depth": 0,
    "index": 0,
    "parent": <root>,
    "siblingsCount": 3,
    "title": "Group Marker",
    "type": "groupMarker",
    "uid": -196255993,
  },
  Object {
    "beta": false,
    "childUIDs": Array [
      -361407047,
      1438225895,
      1439149417,
      1440072939,
    ],
    "depth": 0,
    "groupMarkerUID": -196255993,
    "index": 1,
    "parent": <root>,
    "path": "/foo/child0",
    "siblingsCount": 3,
    "title": "Child0",
    "type": "article",
    "uid": 551503844,
  },
  Object {
    "beta": false,
    "childUIDs": Array [
      1438225895,
      1439149417,
      1440072939,
    ],
    "deprecatedChildrenCount": 0,
    "depth": 1,
    "index": 0,
    "parent": 551503844,
    "siblingsCount": 4,
    "title": "Group Marker, Child 0",
    "type": "groupMarker",
    "uid": -361407047,
  },
  Object {
    "childUIDs": Array [],
    "depth": 1,
    "groupMarkerUID": -361407047,
    "index": 1,
    "parent": 551503844,
    "path": "/foo/child0/grandchild0",
    "siblingsCount": 4,
    "title": "Child0_GrandChild0",
    "type": "tutorial",
    "uid": 1438225895,
  },
  Object {
    "childUIDs": Array [
      305326087,
    ],
    "depth": 1,
    "groupMarkerUID": -361407047,
    "index": 2,
    "parent": 551503844,
    "path": "/foo/child0/grandchild1",
    "siblingsCount": 4,
    "title": "Child0_GrandChild1",
    "type": "tutorial",
    "uid": 1439149417,
  },
  Object {
    "childUIDs": Array [],
    "depth": 2,
    "index": 0,
    "parent": 1439149417,
    "path": "/foo/child0/grandchild0/greatgrandchild0",
    "siblingsCount": 1,
    "title": "Child0_GrandChild0_GreatGrandChild0",
    "type": "tutorial",
    "uid": 305326087,
  },
  Object {
    "childUIDs": Array [],
    "depth": 1,
    "groupMarkerUID": -361407047,
    "index": 3,
    "parent": 551503844,
    "path": "/foo/child0/grandchild2",
    "siblingsCount": 4,
    "title": "Child0_GrandChild2",
    "type": "tutorial",
    "uid": 1440072939,
  },
  Object {
    "childUIDs": Array [
      -827353283,
    ],
    "depth": 0,
    "groupMarkerUID": -196255993,
    "index": 2,
    "parent": <root>,
    "path": "/foo/child1/",
    "siblingsCount": 3,
    "title": "Child1",
    "type": "tutorial",
    "uid": -97593391,
  },
  Object {
    "childUIDs": Array [],
    "depth": 1,
    "index": 0,
    "parent": -97593391,
    "path": "/foo/child1/grandchild0",
    "siblingsCount": 1,
    "title": "Child1_GrandChild0",
    "type": "method",
    "uid": -827353283,
  },
]
`;
